### PR TITLE
support 16.2

### PR DESCRIPTION
**Pull request by venkataanil**
OSP 16.2 needs RHEL 8.4 and OSP 13 needs RHEL 7.9.
Both Scale and Alias forieman has RHEL 8.2 and 7.8
and provided a script to upgrade to desired minor version after
provisioning with foreman.

### CODE REVIEW

File: tasks/install_os.yml [hunk 1]
```diff
@@ -10,25 +10,20 @@
...
   vars:
     rhel_stdout: "{{ (rhel_release.results[0].stdout) | ternary(rhel_release.results[0].stdout, rhel_release.results[1].stdout) }}"
   set_fact:
-    os_install: "{{ needed_os }}"
-  when: rhel_stdout == "" or (needed_os.split()[1] not in rhel_stdout) or (os_search_string not in rhel_stdout)
-
-- name: set os_install before updating os RHEL 7.9
+    os_install: "{{ (needed_os == 'RHEL 8.4') | ternary('RHEL 8.2', ((needed_os == 'RHEL 7.9')| ternary('RHEL 7.8', needed_os))) }}"
+  when: (force_reprovision == true) or (rhel_stdout == "" or (needed_os.split()[1] not in rhel_stdout) or (os_search_string not in rhel_stdout))
+
```
Comment (asyedham) [inline tasks/install_os.yml]: linter is failing due to trailing space on line 22 https://github.com/redhat-performance/jetpack/runs/4332652570?check_suite_focus=true#step:4:97

**Comment by asyedham** [comment]
Deployment is failing
TASK [set hypervisor host to PXE boot off Foreman (Dell)] *************************************************************
Monday 29 November 2021  03:07:27 -0500 (0:00:00.116)       0:07:57.053 ******* 
FAILED - RETRYING: set hypervisor host to PXE boot off Foreman (Dell) (5 retries left).
FAILED - RETRYING: set hypervisor host to PXE boot off Foreman (Dell) (4 retries left).
FAILED - RETRYING: set hypervisor host to PXE boot off Foreman (Dell) (3 retries left).
FAILED - RETRYING: set hypervisor host to PXE boot off Foreman (Dell) (2 retries left).
FAILED - RETRYING: set hypervisor host to PXE boot off Foreman (Dell) (1 retries left).
fatal: [localhost]: FAILED! => {"attempts": 5, "changed": true, "cmd": "source /home/asma/badfish/.venv/bin/activate\np
ython3 badfish.py -H mgmt-e26-h13-740xd.alias.bos.scalelab.redhat.com -u quads -p bos1@878 -i \"/home/asma/badfish/conf
ig/idrac_interfaces.yml\" -t foreman --pxe\n", "delta": "0:00:00.410627", "end": "2021-11-29 03:07:46.142328", "msg": "
non-zero return code", "rc": 1, "start": "2021-11-29 03:07:45.731701", "stderr": "Traceback (most recent call last):\n 
 File \"badfish.py\", line 23, in <module>\n    from helpers.async_lru import alru_cache\nModuleNotFoundError: No modul
e named 'helpers'", "stderr_lines": ["Traceback (most recent call last):", "  File \"badfish.py\", line 23, in <module>
", "    from helpers.async_lru import alru_cache", "ModuleNotFoundError: No module named 'helpers'"], "stdout": "", "st
dout_lines": []}

PLAY RECAP ************************************************************************************************************
localhost                  : ok=49   changed=11   unreachable=0    failed=1    skipped=12   rescued=0    ignored=3

**Comment by venkataanil** [comment]
> Deployment is failing TASK [set hypervisor host to PXE boot off Foreman (Dell)] ************************************************************* Monday 29 November 2021 03:07:27 -0500 (0:00:00.116) 0:07:57.053 ******* FAILED - RETRYING: set hypervisor host to PXE boot off Foreman (Dell) (5 retries left). FAILED - RETRYING: set hypervisor host to PXE boot off Foreman (Dell) (4 retries left). FAILED - RETRYING: set hypervisor host to PXE boot off Foreman (Dell) (3 retries left). FAILED - RETRYING: set hypervisor host to PXE boot off Foreman (Dell) (2 retries left). FAILED - RETRYING: set hypervisor host to PXE boot off Foreman (Dell) (1 retries left). fatal: [localhost]: FAILED! => {"attempts": 5, "changed": true, "cmd": "source /home/asma/badfish/.venv/bin/activate\np ython3 badfish.py -H mgmt-e26-h13-740xd.alias.bos.scalelab.redhat.com -u quads -p bos1@878 -i "/home/asma/badfish/conf ig/idrac_interfaces.yml" -t foreman --pxe\n", "delta": "0:00:00.410627", "end": "2021-11-29 03:07:46.142328", "msg": " non-zero return code", "rc": 1, "start": "2021-11-29 03:07:45.731701", "stderr": "Traceback (most recent call last):\n File "badfish.py", line 23, in \n from helpers.async_lru import alru_cache\nModuleNotFoundError: No modul e named 'helpers'", "stderr_lines": ["Traceback (most recent call last):", " File "badfish.py", line 23, in ", " from helpers.async_lru import alru_cache", "ModuleNotFoundError: No module named 'helpers'"], "stdout": "", "st dout_lines": []}
> 
> PLAY RECAP ************************************************************************************************************ localhost : ok=49 changed=11 unreachable=0 failed=1 skipped=12 rescued=0 ignored=3

Looks like this error message is not related with the current changes.

**Comment by sanjaychari** [comment]
Getting the error 
```
TASK [waiting for the undercloud to be available] ************************************************************************************************
task path: /home/sanjay/jetpack/tasks/wait_after_node_reboot.yml:15
original message: An unhandled exception occurred while templating '
{{ hypervisor_host }}'. Error was a <class 'ansible.errors.AnsibleError'>, original message: recursive loop detected in template string: {{ hyperv
isor_host }}"
```
on scale lab allocation with 5039ms and 6018r systems, with composable roles enabled.

**Comment by sanjaychari** [comment]
> Getting the error
> 
> ```
> TASK [waiting for the undercloud to be available] ************************************************************************************************
> task path: /home/sanjay/jetpack/tasks/wait_after_node_reboot.yml:15
> original message: An unhandled exception occurred while templating '
> {{ hypervisor_host }}'. Error was a <class 'ansible.errors.AnsibleError'>, original message: recursive loop detected in template string: {{ hyperv
> isor_host }}"
> ```
> 
> on scale lab allocation with 5039ms and 6018r systems, with composable roles enabled.

According to this github issue https://github.com/ansible/ansible/issues/8603, I think the issue might be because of this line https://github.com/redhat-performance/jetpack/blob/cf41462f75aee5f2ec8199d3838c4f8881be9675/tasks/install_os.yml#L129.

File: tasks/install_os.yml [hunk 2]
```diff
@@ -10,25 +10,20 @@
...
     rhel_stdout: "{{ (rhel_release.results[0].stdout) | ternary(rhel_release.results[0].stdout, rhel_release.results[1].stdout) }}"
   set_fact:
-    os_install: "{{ needed_os }}"
-  when: rhel_stdout == "" or (needed_os.split()[1] not in rhel_stdout) or (os_search_string not in rhel_stdout)
-
-- name: set os_install before updating os RHEL 7.9
-  set_fact:
-    os_install: "RHEL 7.8"
-  when: needed_os == "RHEL 7.9"
+    os_install: "{{ (needed_os == 'RHEL 8.4') | ternary('RHEL 8.2', ((needed_os == 'RHEL 7.9')| ternary('RHEL 7.8', needed_os))) }}"
```
Comment (asyedham) [inline tasks/install_os.yml]: Setting this fact always re-installs the OS. Even if the RHEL 8.2 is already installed for updating OS to RHEL 8.4

**Review by asyedham — CHANGES_REQUESTED** [review]

File: tasks/install_os.yml [hunk 3]
```diff
@@ -10,25 +10,20 @@
...
+  when: (force_reprovision == true) or (rhel_stdout == "" or (needed_os.split()[1] not in rhel_stdout) or (os_search_string not in rhel_stdout))
 
-- name: set os_install for forceful provisioning
+- name: set upgrade_os
   set_fact:
-    os_install: "{{ needed_os }}"
-  when: force_reprovision == true and needed_os != "RHEL 7.9"
+    upgrade_os: "{{ needed_os.split()[1] }}"
+  when: (needed_os == 'RHEL 8.4') or (needed_os == 'RHEL 7.9')
```
Comment (asyedham) [inline tasks/install_os.yml]: Looks like RHEL 7.9 image is available in both Scale and Alias lab. So we can just upgrade OS for OSP 16.2 to RHEL 8.4

**Review by asyedham — CHANGES_REQUESTED** [review]

File: tasks/install_os.yml [hunk 4]
```diff
@@ -113,3 +108,23 @@
...
+- name: upgrade os and reboot
+  block:
+    - name: upgrade os
+      shell: |
+        sshpass -p {{ hypervisor_password }}  ssh -o 'PreferredAuthentications=password' -o 'StrictHostKeyChecking=no' -o 'UserKnownHostsFile=/dev/null' -q root@{{ hypervisor_host }} '/root/update-latest-rhel-release.sh {{ upgrade_os }} && dnf update -y && reboot'
+      ignore_errors: true
+
+    - include_tasks: tasks/wait_after_node_reboot.yml
+      vars:
+        hypervisor_host: "{{ hypervisor_host }}"
```
Comment (sanjaychari) [inline tasks/install_os.yml]: Why is `hypervisor_host` defined recursively here when it is also defined in https://github.com/redhat-performance/jetpack/blob/cf41462f75aee5f2ec8199d3838c4f8881be9675/prepare_undercloud.yml#L41? `hypervisor_host` variable can be accessed in `tasks/wait_after_node_reboot.yml` without this line.

File: tasks/install_os.yml (hunk 2 above)
Comment (venkataanil) [reply tasks/install_os.yml]: Yes, my intention was code simplicity.
May be I need to fix that.

File: tasks/install_os.yml (hunk 3 above)
Comment (venkataanil) [reply tasks/install_os.yml]: Oh, I was not knowing that.

**Comment by venkataanil** [comment]
> , with composable roles enabled.

Oh, thats a nice finding. Can you please check if we can use hypervisor_host without passing as vars?

**Comment by sanjaychari** [comment]
> > , with composable roles enabled.
> 
> Oh, thats a nice finding. Can you please check if we can use hypervisor_host without passing as vars?

This error does not occur when `hypervisor_host` is not passed as `vars` in https://github.com/redhat-performance/jetpack/blob/cf41462f75aee5f2ec8199d3838c4f8881be9675/tasks/install_os.yml#L129.

**Comment by asyedham** [comment]
LGTM!

**Review by asyedham — APPROVED** [review]
LGTM

**Pull request merged**